File: techniques/two_pointers/twoPointers.cpp
#include <iostream>
#include <string>
#include <vector>

bool isPalindrome(const std::string& word) {
    /*
    Check if word is palindrome.

    Palindrome words can be read same from backward and forward.

    Args:
    word (const std::string& word): The word to check.

    Returns:
    bool: True if word is palindrome else False
    */

    // Initizalize two pointers at start and end, moving toward center
    int leftPointer = 0;
    int rightPointer = word.length() - 1;

    // Compare characters from the start and end, moving toward center
    while (leftPointer < rightPointer) {
        if (word[leftPointer] != word[rightPointer]) {
            return false; // Not a palindrome if characters differ
        }

        leftPointer++; // Move left pointer one step to the right
        rightPointer--; // Move right pointer one step to the left
    }

    return true; // If no mismatches we can assume word is palindrome and return true
}

std::vector<std::vector<int>> twoSum(std::vector<int> &array, int target) {
    int left = 0, right = array.size() - 1;

    std::vector<std::vector<int>> ans;

    while (left < right) {
        int sum = array.at(left) + array.at(right);

        if (sum == target) {
            ans.push_back({ array.at(left), array.at(right) });
            left++;
            right--;
        }
        else if (sum < target) {
            left ++;
        }
        else if (sum > target) {
            right --;
        }
    }

    if (ans.empty()) {
        ans.push_back({-1});
    }


    return ans;
}

int main() {
    std::vector<int> arr = { 0, 1, 2, 3, 4, 5 };
    int target = 5;

    std::vector<std::vector<int>> answer = twoSum(arr, target);

    for (std::vector<int> pairs: answer) {
        std::cout << pairs[0] << " " << pairs[1] << std::endl;
    }

    std::cout << std::endl;

    return 0;
}

File: techniques/two_pointers/two_pointers.py
# Define two different words
word1 = "kayak"
word2 = "light"

# Define a function to check if a word is a palindrome
def is_palindrome(word: str) -> bool:
    # Initialize two pointers, one at the start (left) and one at the end (right) of the word
    left, right = 0, len(word) - 1

    # Loop until the pointers meet in the middle
    while left < right:
        # Check if characters at the left and right pointers are different
        if word[left] != word[right]:
            return False  # If they differ, the word is not a palindrome
        
        # Move the left pointer one step to the right
        left += 1
        # Move the right pointer one step to the left
        right -= 1

    # If we complete the loop without finding mismatches, the word is a palindrome
    return True

# Check if each word is a palindrome and print the result
print(is_palindrome(word=word1))  # Expected output: True, since "kayak" is a palindrome
print(is_palindrome(word=word2))  # Expected output: False, since "light" is not a palindrome

def twoSum(array: list, target: int) -> list:
    left = 0;
    right = len(array) - 1

    while left < right:
        if array[left] + array[right] == target:
            return [array[left], array[right]]
        
        elif array[left] + array[right] < target:
            left -= 1

        elif array[left] + array[right] > target:
            right -= 1
        
    return [-1]

number_list = [1, 2, 3, 4, 5]
target = 5

print(twoSum(array=number_list, target=target))